Add sparse materialisers

gmarkall opened the pull request:
Addition of sparse materialisers and tests. 

toArray is moved up to OGArray in librdag so that it can be accessed by diagonal and sparse matrices.

Buildbot pass: http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/362

Librdag coverage: 99.4% (needs another commit to get up to 100%)
jshim coverage: 20.7%

gmarkall committed:
[MAT-261] Add materialise for real sparse matrix.

gmarkall committed:
[MAT-261] Add materialise of complex sparse matrix and tests from Java side.

gmarkall committed:
[MAT-261] Fix build on Windows.

It's always Windows.

gmarkall committed:
[MAT-261] Add testing of sparse conversions to array.

gmarkall commented:
Rebased and added missing rdag tests.

Buildbot pass: http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/364

Librdag coverage: 88.6% (same as master)
jshim coverage: 20.6% (slightly lower than master, due to adding some untested functions in a set of similar untested functions)

stuartarchibald commented on src/librdag/terminal.cc:
```diff
@@ -197,6 +197,18 @@
+  {
+    tmp[i] = _data[i];
+  }
+  return tmp;
```
would `memcpy (tmp, _data, sizeof(T)*_datalen );` work?

gmarkall replied:
probably. I'll try it and do so if it does. If it does, the memcpy could also probably be used for part of toArrayOfArrays.

gmarkall committed:
[MAT-261] Improve sparse matrix testing

Add more complex sparse matrices for materialise tests.

Correct bounds in RDAG complex sparse matrix test.

Correct colPtr in materialisation of sparse matrices.

gmarkall committed:
[MAT-261] Use memcpy in toArray()

gmarkall commented:
I've added a more complicated sparse matrix and fixed the bug that it exposed. Also, `toArray` is changed to use a `memcpy`.

Buildbot pass: http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/368

stuartarchibald commented:
Happy with this. Couple of good catches enabled by data with different sizes. Memcpy is valuable as our data is flat :)
